executor/utils.go

```diff
@@ -53,3 +58,201 @@ func deleteFromSet(set []string, value string) []string {
+	// queried, the second way will be chosen; otherwise, the first way will be chosen.
+	fetchAllLimit int
+}
+
+// NewSQLDigestTextRetriever creates a new SQLDigestTextRetriever.
+func NewSQLDigestTextRetriever() *SQLDigestTextRetriever {
+	return &SQLDigestTextRetriever{
+		SQLDigestsMap: make(map[string]string),
+		fetchAllLimit: 512,
+	}
+}
+
+func (r *SQLDigestTextRetriever) runMockQuery(data map[string]string, inValues []interface{}) (map[string]string, error) {
+	if len(inValues) == 0 {
+		return data, nil
+	}
+	res := make(map[string]string, len(inValues))
+	for _, digest := range inValues {
+		if text, ok := data[digest.(string)]; ok {
+			res[digest.(string)] = text
+		}
+	}
+	return res, nil
+}
+
+// runQuery runs query to the system tables to fetch normalized SQL texts corresponding to given SQL digests, if
+// `inValues` is given, or all digest -> SQL mappings otherwise. If `queryGlobal` is false, it queries
+// information_schema.statements_summary and information_schema.statements_summary_history; otherwise, it queries the
+// cluster version of these two tables.
+func (r *SQLDigestTextRetriever) runQuery(ctx context.Context, sctx sessionctx.Context, queryGlobal bool, inValues []interface{}) (map[string]string, error) {
+	// If mock data is set, query the mock data instead of the real statements_summary tables.
+	if !queryGlobal && r.mockLocalData != nil {
+		return r.runMockQuery(r.mockLocalData, inValues)
+	} else if queryGlobal && r.mockGlobalData != nil {
+		return r.runMockQuery(r.mockGlobalData, inValues)
+	}
+
+	exec, ok := sctx.(sqlexec.RestrictedSQLExecutor)
+	if !ok {
+		return nil, errors.New("restricted sql can't be executed in this context")
+	}
+
+	// Information in statements_summary will be periodically moved to statements_summary_history. Union them together
+	// to avoid missing information when statements_summary is just cleared.
+	stmt := "select digest, digest_text from information_schema.statements_summary union distinct " +
+		"select digest, digest_text from information_schema.statements_summary_history"
+	if queryGlobal {
+		stmt = "select digest, digest_text from information_schema.cluster_statements_summary union distinct " +
+			"select digest, digest_text from information_schema.cluster_statements_summary_history"
+	}
+	// Add the where clause if `inValues` is specified.
+	if len(inValues) > 0 {
+		stmt += " where digest in (" + strings.Repeat("%?,", len(inValues)-1) + "%?)"
+	}
+
+	stmtNode, err := exec.ParseWithParams(ctx, stmt, inValues...)
+	if err != nil {
+		return nil, err
+	}
+	rows, _, err := exec.ExecRestrictedStmt(ctx, stmtNode)
+	if err != nil {
+		return nil, err
+	}
+
+	res := make(map[string]string, len(rows))
+	for _, row := range rows {
+		res[row.GetString(0)] = row.GetString(1)
+	}
+	return res, nil
+}
+
+func (r *SQLDigestTextRetriever) updateWithQueryResult(queryResult map[string]string) {
```

Comment:
How about `updateDigestInfo`. Or we could add some comments about the targets of the update.